### PR TITLE
macros: implement FromGlibNone/Full for shared boxed types.

### PR DESCRIPTION
**Pull request by pbor**

### CODE REVIEW

File: glib-macros/src/gboxed_shared_derive.rs [hunk 1]
```diff
@@ -178,5 +178,20 @@ pub fn impl_gshared_boxed(input: &syn::DeriveInput) -> proc_macro2::TokenStream
         #impl_to_value_optional
 
         #impl_from_value
+
+        impl #crate_ident::translate::FromGlibPtrNone<*const #refcounted_type_prefix::InnerType> for #name {
```
Comment (sdroege) [inline glib-macros/src/gboxed_shared_derive.rs:182]: For completeness this should probably also be done for
 - `ToGlibPtr`
 - the non-shared `GBoxed` (`ToGlibPtr`, `FromGlibPtrXXX`, `ToGlibPtrMut`)

**Comment by sdroege** [comment]
Looks good to me otherwise

File: glib-macros/src/gboxed_shared_derive.rs (hunk 1 above)
Comment (pbor) [reply glib-macros/src/gboxed_shared_derive.rs:182]: Hehe... yeah, I saw that one coming, but I was not totally sure how to do it for GBoxed, so I went ahead and submitted the pr :)

Comment (sdroege) [reply glib-macros/src/gboxed_shared_derive.rs:182]: The same way how it works in `FromValue` / `ToValue` :) What problem did you run into?

Comment (pbor) [reply glib-macros/src/gboxed_shared_derive.rs:182]: mind you I have not tried very hard... but the first question I run into is what types to use in the ffi... for Shared I can used the `InnerType` associated type

Comment (sdroege) [reply glib-macros/src/gboxed_shared_derive.rs:182]: The type itself, as that's what is also stored as pointer value in the `GValue`

**Comment by pbor** [comment]
Please triple check, I have no idea what I am doing :)

File: glib-macros/src/gboxed_shared_derive.rs [hunk 2]
```diff
@@ -178,5 +178,40 @@ pub fn impl_gshared_boxed(input: &syn::DeriveInput) -> proc_macro2::TokenStream
...
+                unsafe {
+                    #crate_ident::translate::Stash(#refcounted_type_prefix::into_raw(r), self)
+                }
+            }
+
+            #[inline]
+            fn to_glib_full(&self) -> *const #refcounted_type_prefix::InnerType {
+                let r = <#name as #crate_ident::subclass::shared::SharedType>::into_refcounted(self.clone());
+                unsafe {
+                    #refcounted_type_prefix::into_raw(r)
```
Comment (sdroege) [inline glib-macros/src/gboxed_shared_derive.rs:212]:
```suggestion
                    #refcounted_type_prefix::ref_(#refcounted_type_prefix::into_raw(r))
```

Maybe? Not entirely sure, either here that or in `to_glib_none()` it shouldn't use `into_raw()`. The way how it is now it is asymmetric.

File: glib-macros/src/gboxed_derive.rs [hunk 3]
```diff
@@ -162,5 +162,37 @@ pub fn impl_gboxed(input: &syn::DeriveInput) -> TokenStream {
...
 
         #impl_from_value
+
+        impl #crate_ident::translate::FromGlibPtrNone<*const #name> for #name {
+            #[inline]
+            unsafe fn from_glib_none(ptr: *const #name) -> Self {
+                let t = #name::static_type();
+                let ptr = #crate_ident::gobject_ffi::g_boxed_copy(<#crate_ident::Type as #crate_ident::translate::IntoGlib>::into_glib(t),  ptr as *mut _) as *mut #name;
+                assert!(!ptr.is_null());
+                *Box::from_raw(ptr)
```
Comment (sdroege) [inline glib-macros/src/gboxed_derive.rs:172]:
```suggestion
                assert!(!ptr.is_null());
                (&*ptr).clone()
```

File: glib-macros/src/gboxed_derive.rs [hunk 4]
```diff
@@ -162,5 +162,37 @@ pub fn impl_gboxed(input: &syn::DeriveInput) -> TokenStream {
...
+            }
+        }
+
+        impl<'a> #crate_ident::translate::ToGlibPtr<'a, *const #name> for #name {
+            type Storage = &'a Self;
+
+            #[inline]
+            fn to_glib_none(&'a self) -> #crate_ident::translate::Stash<'a, *const #name, Self> {
+                #crate_ident::translate::Stash(Box::into_raw(Box::new(self.clone())), self)
```
Comment (sdroege) [inline glib-macros/src/gboxed_derive.rs:189]:
```suggestion
                #crate_ident::translate::Stash(&self as *const #name, self)
```

File: glib-macros/src/gboxed_derive.rs [hunk 5]
```diff
@@ -162,5 +162,37 @@ pub fn impl_gboxed(input: &syn::DeriveInput) -> TokenStream {
         #impl_to_value_optional
 
         #impl_from_value
+
+        impl #crate_ident::translate::FromGlibPtrNone<*const #name> for #name {
```
Comment (sdroege) [inline glib-macros/src/gboxed_derive.rs:166]: Also with `*mut #name` (can just call into the `const` one)

File: glib-macros/src/gboxed_derive.rs [hunk 6]
```diff
@@ -162,5 +162,37 @@ pub fn impl_gboxed(input: &syn::DeriveInput) -> TokenStream {
...
+
+        impl #crate_ident::translate::FromGlibPtrFull<*mut #name> for #name {
+            #[inline]
+            unsafe fn from_glib_full(ptr: *mut #name) -> Self {
+                assert!(!ptr.is_null());
+                *Box::from_raw(ptr)
+            }
+        }
+
+        impl<'a> #crate_ident::translate::ToGlibPtr<'a, *const #name> for #name {
```
Comment (sdroege) [inline glib-macros/src/gboxed_derive.rs:184]: Also for `*mut #name` (can just call into the `const` one)

File: glib-macros/src/gboxed_shared_derive.rs [hunk 7]
```diff
@@ -178,5 +178,40 @@ pub fn impl_gshared_boxed(input: &syn::DeriveInput) -> proc_macro2::TokenStream
         #impl_to_value_optional
 
         #impl_from_value
+
+        impl #crate_ident::translate::FromGlibPtrNone<*const #refcounted_type_prefix::InnerType> for #name {
+            #[inline]
```
Comment (sdroege) [inline glib-macros/src/gboxed_shared_derive.rs:183]: Also for *mut #name (can just call into the const one)

File: glib-macros/src/gboxed_shared_derive.rs [hunk 8]
```diff
@@ -178,5 +178,40 @@ pub fn impl_gshared_boxed(input: &syn::DeriveInput) -> proc_macro2::TokenStream
...
+        }
+
+        impl #crate_ident::translate::FromGlibPtrFull<*mut #refcounted_type_prefix::InnerType> for #name {
+            #[inline]
+            unsafe fn from_glib_full(ptr: *mut #refcounted_type_prefix::InnerType) -> Self {
+                #name(#refcounted_type_prefix::from_raw(ptr))
+            }
+        }
+
+        impl<'a> #crate_ident::translate::ToGlibPtr<'a, *const #refcounted_type_prefix::InnerType> for #name {
```
Comment (sdroege) [inline glib-macros/src/gboxed_shared_derive.rs:197]: Also for *mut #name (can just call into the const one)

File: glib-macros/src/gboxed_shared_derive.rs [hunk 9]
```diff
@@ -178,5 +178,40 @@ pub fn impl_gshared_boxed(input: &syn::DeriveInput) -> proc_macro2::TokenStream
...
+            type Storage = &'a Self;
+
+            #[inline]
+            fn to_glib_none(&'a self) -> #crate_ident::translate::Stash<'a, *const #refcounted_type_prefix::InnerType, Self> {
+                let r = <#name as #crate_ident::subclass::shared::SharedType>::into_refcounted(self.clone());
+                unsafe {
+                    #crate_ident::translate::Stash(#refcounted_type_prefix::into_raw(r), self)
```
Comment (sdroege) [inline glib-macros/src/gboxed_shared_derive.rs:204]: I think this one is the wrong one. It shouldn't `into_raw()` but just get a pointer to the existing one

File: glib-macros/src/gboxed_shared_derive.rs (hunk 2 above)
Comment (sdroege) [reply glib-macros/src/gboxed_shared_derive.rs:212]: And this one has to be like in my suggestion

**Comment by sdroege** [comment]
> Please triple check, I have no idea what I am doing :)

Done :)

File: glib-macros/src/gboxed_shared_derive.rs (hunk 2 above)
Comment (pbor) [reply glib-macros/src/gboxed_shared_derive.rs:218]: as discussed in chat, we are cloning in the line above

**Comment by sdroege** [comment]
@pbor Wants a `cargo fmt` run

**Review by sdroege — APPROVED** [review]

**Review by sdroege — APPROVED** [review]

**Review by GuillaumeGomez — APPROVED** [review]

**Pull request merged**